### PR TITLE
Header to reagent

### PR DESCRIPTION
**Pull request by 10plusY**
addresses header section of https://app.asana.com/0/106998764700389/110645114924616

Issues: 

- subscribed tags pane doesn't update asynchronously when you unsubscribe to a tag
- not sure if online users pane updates when another use logs off as trying to logging off gets a 403 response

### CODE REVIEW

**Comment by jamesnvc** [comment]
Typos in `src/braid/common/state.cljs`, lines 116, 120, I think you want `user-id` instead of `user_id`

**Comment by jamesnvc** [comment]
The logout thing is happening in master too, I will investigate that

**Comment by 10plusY** [comment]
lol that's why the status subscription wasn't working

**Comment by jamesnvc** [comment]
Looks good; can you delete the now-unused om files & remove the requires?

**Comment by jamesnvc** [comment]
I also kind of want to replace all the uses of `chat.client.views.pills` with `braid.ui.views.pills`, just 'cause it's kind of confusing to have two implementations of those views. @rafd thoughts on doing that in this pull request versus doing that in a new one?

**Comment by 10plusY** [comment]
I removed the om versions of the rest of the header buttons. I didn't want to delete any files until the full transition to reagent has been made

**Comment by 10plusY** [comment]
So for now I could just delete the om versions in chat.client.views.pills so its less ambiguous

**Comment by 10plusY** [comment]
but yeah will remove requires etc

**Comment by jamesnvc** [comment]
I think you can go ahead and delete `src/chat/client/views/header.cljs`, as that doesn't seem to be used anywhere.
I'm just hesitant about the pills stuff because I think there are still a bunch of places that are using the om pills views, so all those files would have to change too, hence maybe thinking it should be a separate pull request...

**Comment by 10plusY** [comment]
of course. yah I wasn't going to delete the om pill views until all of the other components used the reagent version of the pills.

**Comment by rafd** [comment]
re: pills, we can do it in stages, in a separate PR:
 1)  wrap the reagent pill views with an om component (so as not to change anywhere else)
 2) change all the places that use the om pills view to use the reagent one (and delete the om one)

**Comment by rafd** [comment]
@10plusY, re issues:

2) does it work when a new user logs on?

File: src/braid/common/state.cljs [hunk 1]
```diff
@@ -62,3 +62,65 @@
...
+                                (or (empty? (thread :tag-ids))
+                                    (contains?
+                                      (into #{} (map group-for-tag (thread :tag-ids)))
+                                      @current-group-id))))))]
+      (reaction open-threads)))
+
+
+(defn- users-in-group
+  [group-id state]
+  (filter (fn [u] (contains? (set (u :group-ids)) group-id)) (vals (state :users))))
```
Comment (rafd) [inline src/braid/common/state.cljs]: might as well make this act like a subscription. ie take in [state [_ group-id] and return a reaction (will need to change places that use it)

File: src/braid/common/state.cljs [hunk 2]
```diff
@@ -62,3 +62,65 @@
...
+
+(defn get-group-subscribed-tags
+  [state [_ group-id]]
+  (reaction
+    (->> (vals (get-in @state [:tags]))
+      (filter (fn [tag] (= (get-in @state [:open-group-id]) (tag :group-id))))
+      (filter (fn [tag] @(get-user-subscribed-to-tag @state [nil (tag :id)]))))))
+
+(defn get-user-avatar-url
+  [state user-id]
```
Comment (rafd) [inline src/braid/common/state.cljs]: this should probably be: [state [_ user-id]]

File: src/braid/common/state.cljs [hunk 3]
```diff
@@ -62,3 +62,65 @@
...
+  (reaction (get-in @state [:users user-id :avatar])))
+
+(defn get-user-status
+  [state user-id]
```
Comment (rafd) [inline src/braid/common/state.cljs]: this should probably be: [state [_ user-id]]

File: src/braid/common/state.cljs [hunk 4]
```diff
@@ -62,3 +62,65 @@
...
+  [state _]
+  (let [current-group-id (reaction (@state :open-group-id))
+        open-thread-ids (reaction (get-in @state [:user :open-thread-ids]))
+        group-for-tag (fn [tag-id]
+                        (get-in @state [:tags tag-id :group-id]))
+        threads (reaction (@state :threads))
+        open-threads (-> @threads
+                         (select-keys @open-thread-ids)
+                         vals
+                         #_(->> (filter (fn [thread]
```
Comment (rafd) [inline src/braid/common/state.cljs]: this filter step is commented out. may need to review if it's necessary.

File: src/braid/ui/views/header.cljs [hunk 5]
```diff
@@ -0,0 +1,113 @@
+(ns braid.ui.views.header
+  (:require [chat.client.dispatcher :refer [dispatch!]]
+            [chat.client.routes :as routes]
+            [chat.client.views.helpers :refer [id->color]]
+            [braid.ui.views.pills :refer [tag-pill-view user-pill-view]]
+            [braid.ui.views.search-bar :refer [search-bar-view]]
+            [om.core :as om]))
```
Comment (rafd) [inline src/braid/ui/views/header.cljs]: review which requires are needed. om.core probably isn't.

File: src/braid/ui/views/header.cljs [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    (fn []
+      [:div
+        (when (< 5 (count @open-threads))
+          [:button {:on-click (fn [_]
+                                (dispatch! :clear-inbox))}
+            "Clear Inbox"])])))
+
+(defn inbox-page-button-view []
+  (let [path (routes/inbox-page-path {:group-id (routes/current-group)})]
+    [:div.inbox.shortcut {:className (when (routes/current-path? path) "active")}
```
Comment (rafd) [inline src/braid/ui/views/header.cljs]: use `:class` instead of `:className`

File: src/braid/ui/views/header.cljs [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+      [:a {:href path
+           :className "title"
+           :title "Inbox"}]]))
+
+(defn recent-page-button-view []
+  (let [path (routes/recent-page-path {:group-id (routes/current-group)})]
```
Comment (rafd) [inline src/braid/ui/views/header.cljs]: `routes/current-group` accesses `@store/app-state` directly. will need to use a `subscribe` instead

File: src/braid/ui/views/header.cljs [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+      [:div.modal
+        [:p "Conversations must be tagged to be seen by other people."]
+        [:p "Tag a conversation by mentioning a tag in a message: ex. #general"]
+        [:p "You can also mention users to add them to a conversation: ex. @raf"]
+        [:p "Add emoji by using :shortcodes: (they autocomplete)."]]]))
+
+(defn users-online-pane-view [{:keys [subscribe]}]
+  (let [user-id (subscribe [:user-id])
+        users-in-open-group (subscribe [:users-in-open-group])
+        users (->> @users-in-open-group
```
Comment (rafd) [inline src/braid/ui/views/header.cljs]: I think all derefs must happen within `(fn [] ...`

File: src/braid/ui/views/header.cljs [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        (for [user users-online]
+          [user-pill-view user subscribe])]])))
+
+
+(defn tags-pane-view [{:keys [subscribe]}]
+  (let [path (routes/page-path {:group-id (routes/current-group)
+                                :page-id "channels"})
+        tags (subscribe [:tags])
+        group-subscribed-tags (subscribe [:group-subscribed-tags])]
+    [:div.tags.shortcut {:className (when (routes/current-path? path) "active")}
```
Comment (rafd) [inline src/braid/ui/views/header.cljs]: missing an internal `(fn [] ...`

File: src/braid/ui/views/header.cljs [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+(defn current-user-button-view [{:keys [subscribe]}]
+  (let [user-id (subscribe [:user-id])
+        user-avatar-url (subscribe [:user-avatar-url])
+        path (routes/page-path {:group-id (routes/current-group)
+                                :page-id "me"})]
+    (fn []
+      [:a {:href path
+           :className (when (routes/current-path? path) "active")}
+        [:img.avatar {:style {:backgroundColor (id->color @user-id)}
```
Comment (rafd) [inline src/braid/ui/views/header.cljs]: use `:background-color`

File: src/braid/ui/views/pills.cljs [hunk 11]
```diff
@@ -0,0 +1,38 @@
+(ns braid.ui.views.pills
+  (:require [om.core :as om]
+            [chat.client.routes :as routes]
+            [chat.client.views.helpers :refer [id->color user-cursor]]))
+
+(defn tag-pill-view
+  [tag subscribe]
+  (let [path (routes/tag-page-path {:group-id (routes/current-group)
+                                    :tag-id (tag :id)})
```
Comment (rafd) [inline src/braid/ui/views/pills.cljs]: as noted in header.cljs, will need to use a `subscribe` to replace `routes/current-group` and bind `path *inside* of the inner render `fn`

File: src/braid/ui/views/search_bar.cljs [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+          (go
+            (while true
+              (let [{:keys [query]} (<! search)]
+                (store/set-search-results! {})
+                (if (string/blank? query)
+                  (do
+                    (routes/go-to! (routes/inbox-page-path {:group-id (routes/current-group)})))
+                  (do
+                    (routes/go-to! (routes/search-page-path {:group-id (routes/current-group)
+                                                             :query query}))
```
Comment (rafd) [inline src/braid/ui/views/search_bar.cljs]: will need to create a `subscribe` to replace `routes/current-group`

File: src/chat/client/views.cljs [hunk 13]
```diff
@@ -76,7 +79,9 @@
 
          (SidebarView. #js {:subscribe subscribe})
 
-        (om/build header-view data)
+        (HeaderView. #js {:subscribe subscribe})
+
+        ;(om/build chat.client.views.header/header-view data)
```
Comment (rafd) [inline src/chat/client/views.cljs]: get rid of the commented-out code

File: src/braid/ui/views/search_bar.cljs [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+         [:div.search-bar
+           [:input {:type "search"
+                    :placeholder "Search History"
+                    :value @search-query
+                    :on-change
+                      (fn [e]
+                        (let [query (.. e -target -value)]
+                          (store/set-search-query! query)
+                          (put! search-chan {:query query})))}]])})))
+
```
Comment (rafd) [inline src/braid/ui/views/search_bar.cljs]: remote the old search_bar code in `chat/client/views/search_bar.cljs`

File: src/braid/ui/views/pills.cljs [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+    (fn []
+      (let [path (routes/user-page-path {:group-id @open-group-id
+                                         :user-id (user :id)})]
+        [:a.user.pill {:class (case @user-status
+                                              :online " on"
+                                                      " off")
+                       :tabIndex -1
+                       :style {:background-color (id->color (user :id))
+                               :color (id->color (user :id))
+                               :borderColor (id->color (user :id))}
```
Comment (rafd) [inline src/braid/ui/views/pills.cljs]: use `:border-color`

File: src/braid/ui/views/pills.cljs [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+    (fn []
+      (let [path (routes/tag-page-path {:group-id @open-group-id
+                                        :tag-id (tag :id)})]
+        [:a.tag.pill {:class (if @user-subscribed-to-tag?
+                                " on"
+                                " off")
+                      :tabIndex -1
+                      :style {:background-color (id->color (tag :id))
+                              :color (id->color (tag :id))
+                              :borderColor (id->color (tag :id))}
```
Comment (rafd) [inline src/braid/ui/views/pills.cljs]: use `:border-color`

File: src/braid/ui/views/header.cljs [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+                                (dispatch! :clear-inbox))}
+            "Clear Inbox"])])))
+
+(defn inbox-page-button-view [{:keys [subscribe]}]
+  (let [open-group-id (subscribe [:open-group-id])]
+    (fn []
+      (let [path (routes/inbox-page-path {:group-id @open-group-id})]
+        [:div.inbox.shortcut {:class (when (routes/current-path? path) "active")}
+          [:a {:href path
+               :class "title"
```
Comment (rafd) [inline src/braid/ui/views/header.cljs]: can do `:a.title` (and in other places)

File: src/braid/ui/views/pills.cljs [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+(defn tag-pill-view
+  [tag subscribe]
+  (let [open-group-id (subscribe [:open-group-id])
+        user-subscribed-to-tag? (subscribe [:user-subscribed-to-tag (tag :id)])]
+    (fn []
+      (let [path (routes/tag-page-path {:group-id @open-group-id
+                                        :tag-id (tag :id)})]
+        [:a.tag.pill {:class (if @user-subscribed-to-tag?
+                                " on"
+                                " off")
```
Comment (rafd) [inline src/braid/ui/views/pills.cljs]: no need for space in the strings anymore, can also do on a single line
ie:
`(if @user-subscribed-to-tag? "on" "off")`

**Comment by rafd** [comment]
added two more tiny comments.

can you confirm that you tested and all the header stuff seems to be working?

**Comment by 10plusY** [comment]
The search bar works fine but the subscribed tags pane doesn't update automatically when I unsubscribe

**Pull request merged**